### ISSUE TITLE
Bug in Parent constraint

### ISSUE THREAD

**Issue by Soterr8** [Rainbox-dev/DuAEF_Duik#683]
**Describe the bug**
Parent constraint function doesn't work properly and fails to establish 100% weight following after the 18-th shot. The parent constraint works till the 17-th shot and then starts to fail.

**To Reproduce**
1. Create two layers (parent and child layers).
2. Parent the child layer with parent constraint to the patent layer.
3. Set weight of parent constraint to 100%.
4. Create a keyframe on parent layer on 0 frame.
5. Move playhead forward till 18-th shot and change the position of the parent layer. The child layer doesn't implement 100% weight following since this shot.
**Error description, if any**
File: 
Line number:
Error message: 

**Expected behavior**
The child layer is expected to follow the parent layer for 100% without offset

**Screenshots**

**System (please complete the following information):**
 - OS: Windows 10
 - Version of After Effects: 17.0.14
 - Version of Duik: 16.2.11

**Additional context**
Thank you for your cript!! I would appreciate your answering!!
[Project_and_screenshots.zip](https://github.com/Rainbox-dev/DuAEF_Duik/files/4652486/Project_and_screenshots.zip)

**Comment by mamoun86** [comment]
or Automations > Move Away

**Comment by Nico-Duduf** [comment]
Hi

thanks for your feedback.
What do you mean exactly by 18th shot? What are you calling "shot"? You mean frame/time in the composition?

**Comment by Soterr8** [comment]
Hello, I mean that the trouble occurs starting from the 18th frame. I have tried on different computers, but the result is the same. However, I have used parent constraint across compositions and position constraint for the same composition and it worked properly.

**Comment by Soterr8** [comment]
I spent two days on it, but I can not find a solution

**Comment by Nico-Duduf** [comment]
Thanks for your feedback again, you've found a bug in After Effects ;)
But I've been able to implement a workaround in Duik, it will be available soon in the next update, 16.2.12.

In the meantime if you want to fix your project, replace the expression in the position of the child by this one:
```js
//Duik.parentConstraint
function translatePointWithLayer( l, point, startT, endT ) {
if (typeof startT === "undefined") startT = 0;
if (typeof endT === "undefined") endT = time;
try {
var pos = l.fromWorld( point, startT );
} catch ( e ) {
var pos = [ 0, 0 ];
}
var prevPos = l.toWorld( pos, startT );
var newPos = l.toWorld( pos, endT );
return newPos - prevPos;
}
function checkDuikEffect(fx, duikMatchName) {
if (fx.numProperties  < 3) return false;
if (!!$.engineName) {
if ( fx(2).name != duikMatchName ) return false;
}
else {
try { if (fx(2).name != duikMatchName) return false; }
catch (e) { return false; }
}
return true;
}
var result = thisLayer.position.valueAtTime( 0 );
var mbPrecision = 1;
var step = framesToTime(1);
var cT = 0;
while ( cT < time ) {
    cT += step;
    var roundedFrame = Math.round( timeToFrames(cT) );
    var roundedTime = roundedFrame * thisComp.frameDuration;
    if (cT - roundedTime < 0.0001) cT = roundedTime;

	for ( var i = 1; i <= thisLayer( "Effects" ).numProperties; i++ ) {
		var fx = effect( i );
		if ( !checkDuikEffect(fx, "DUIK parentConstraint2") ) continue;
		if ( !fx(7).value ) continue;
		var parentLayer = null;
		try {
			parentLayer = fx(4);
		} catch ( e ) {
			continue;
		}

		mbPrecision = fx(10).value;
		step = framesToTime(1)/mbPrecision;

        if ( !parentLayer ) continue;
        if ( parentLayer.index == index ) continue;
        var weight = fx(5).valueAtTime( cT ) / 100;
        if ( weight == 0 ) continue;
        result += translatePointWithLayer( parentLayer, result, cT - step, cT ) * weight;
    }
    result += valueAtTime( cT ) - valueAtTime( cT - step );
}
result;
```

**Comment by Soterr8** [comment]
Thank you very much for solution and for your work!!
